### PR TITLE
[SYCL][ESIMD] Disable bitreverse recognition for non-std width

### PR DESCRIPTION
**Pull request by v-klochkov**
SPIR targets require SPV_KHR_bit_instructions and
SPV_INTEL_arbitrary_precision_integers extensions to translate llvm.bitreverse with non-standard width and this combination is not yet supported by SPV translator and/or device back-ends.

### CODE REVIEW

**Comment by maksimsab** [comment]
I would propose to add LIT test that checks one of LLVM Transformation passes.

File: llvm/lib/Transforms/Utils/Local.cpp [hunk 1]
```diff
@@ -3993,6 +3993,14 @@ bool llvm::recognizeBSwapOrBitReverseIdiom(
   if (DemandedBW > ITy->getScalarSizeInBits())
     return false;
 
+  // SPIR targets require SPV_KHR_bit_instructions and
+  // SPV_INTEL_arbitrary_precision_integers extensions to translate
+  // llvm.bitreverse with non-standard width and this combination is not yet
+  // supported by SPV translator and/or device back-ends.
+  if (Triple(I->getModule()->getTargetTriple()).isSPIR() &&
+      I->getModule()->getDataLayout().isIllegalInteger(DemandedBW))
```
Comment (sarnex) [inline llvm/lib/Transforms/Utils/Local.cpp:4001]: How difficult is it to add support to the translator?

IGC used the upstream translator now, so if we fix it there they should be able to reverse-translate this new pattern *relatively* quickly

Comment (LU-JOHN) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: > How difficult is it to add support to the translator?

The translator already supports SPV_INTEL_arbitrary_precision_integers.  This is not enabled by the driver by default because it was only intended for FPGA and emulation https://github.com/intel/llvm/pull/13379 .
The translator already supports SPV_KHR_bit_instructions. This is not enabled in SYCLOS because backends do not support it:  https://github.com/intel/llvm/pull/13044 .  It is enabled as an Intel customization:  https://github.com/intel-restricted/applications.compilers.llvm-project/pull/20392 

To summarize the translator can represent llvm.bitreverse.* calls with SPIR-V BitReverse for non-standard sizes if SPV_INTEL_arbitrary_precision_integers and SPV_KHR_bit_instructions are enabled.

Comment (sarnex) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Ok, thanks

Comment (v-klochkov) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: It takes too much time for driver to get updated and reach our CI and customers.
I suggest to proceed with disabling non-std width bitreverse, let driver enable support by default, wait for that driver to reach CI, then remove the lines added in this patch.

Comment (sarnex) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Sure

Comment (LU-JOHN) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Also, the translator will lower llvm.bitreverse.* calls to basic operations if SPV_KHR_bit_instructions is not enabled.  Support for 2/4-bit widths was added in https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2481.  What non-std width was being used?

Comment (LU-JOHN) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Can we add:

> !(SPV_INTEL_arbitrary_precision_integers &&
>   SPV_KHR_bit_instructions)

to the condition and make this a permanent change?  This will allow the user to control if llvm.bitreverse.* is generated for non-standard widths based on the supported extensions.

Comment (v-klochkov) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: I don't mind adding it. Can you advise on how to do such check/query here in InstCombine/Local.cpp?

Comment (LU-JOHN) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Testing for SPV_* extensions can be done with SPIRVSubtarget::canUseExtension, but is only done within llvm/lib/Target/SPIRV.  As far as I can tell, transform passes cannot or should not access this extension info.

**Review by bso-intel — APPROVED** [review]

File: llvm/lib/Transforms/Utils/Local.cpp (hunk 1 above)
Comment (v-klochkov) [reply llvm/lib/Transforms/Utils/Local.cpp:4001]: Thank you for the comments here.
I would add that both extensions (`SPV_KHR_bit_instructions` and `SPV_INTEL_arbitrary_precision_integers`)
are already specified in the arguments of LLVM-to-SPIRV conversion:
```
"llvm-foreach" --in-file-list=/tmp/vadd_usm-aa8f3b.txt --in-replace=/tmp/vadd_usm-aa8f3b.txt --out-ext=spv --out-file-list=/tmp/vadd_usm-6e4b2e.txt --out-replace=/tmp/vadd_usm-6e4b2e.txt -- build/install/bin/llvm-spirv -o /tmp/vadd_usm-6e4b2e.txt -spirv-max-version=1.4 -spirv-debug-info-version=nonsemantic-shader-200 -spirv-allow-unknown-intrinsics=llvm.genx. -spirv-ext=-all,+SPV_EXT_shader_atomic_float_add,+SPV_EXT_shader_atomic_float_min_max,+SPV_KHR_no_integer_wrap_decoration,+SPV_KHR_float_controls,+SPV_KHR_expect_assume,+SPV_KHR_linkonce_odr,+SPV_INTEL_subgroups,+SPV_INTEL_media_block_io,+SPV_INTEL_device_side_avc_motion_estimation,+SPV_INTEL_fpga_loop_controls,+SPV_INTEL_unstructured_loop_controls,+SPV_INTEL_fpga_reg,+SPV_INTEL_blocking_pipes,+SPV_INTEL_function_pointers,+SPV_INTEL_kernel_attributes,+SPV_INTEL_io_pipes,+SPV_INTEL_inline_assembly,+SPV_INTEL_arbitrary_precision_integers,+SPV_INTEL_float_controls2,+SPV_INTEL_vector_compute,+SPV_INTEL_fast_composite,+SPV_INTEL_arbitrary_precision_fixed_point,+SPV_INTEL_arbitrary_precision_floating_point,+SPV_INTEL_variable_length_array,+SPV_INTEL_fp_fast_math_mode,+SPV_INTEL_long_constant_composite,+SPV_INTEL_arithmetic_fence,+SPV_INTEL_global_variable_decorations,+SPV_INTEL_cache_controls,+SPV_INTEL_fpga_buffer_location,+SPV_INTEL_fpga_argument_interfaces,+SPV_INTEL_fpga_invocation_pipelining_attributes,+SPV_INTEL_fpga_latency_control,+SPV_INTEL_task_sequence,+SPV_INTEL_token_type,+SPV_INTEL_bfloat16_conversion,+SPV_INTEL_joint_matrix,+SPV_INTEL_hw_thread_queries,+SPV_KHR_uniform_group_instructions,+SPV_INTEL_masked_gather_scatter,+SPV_INTEL_tensor_float32_conversion,+SPV_INTEL_optnone,+SPV_KHR_non_semantic_info /tmp/vadd_usm-aa8f3b.txt
```
but SPIRV-gen phase does not replace them with some emulation sequences and keeps operations like this (see SPIRV converted back to LLVM-IR):
```
  %trunc.i.i = trunc i32 %0 to i2
  %rev.i.i = call spir_func i2 @llvm_bitreverse_i2(i2 %trunc.i.i) #1
```

Comment (LU-JOHN) [reply llvm/lib/Transforms/Utils/Local.cpp:4045]: > I would add that both extensions (SPV_KHR_bit_instructions and SPV_INTEL_arbitrary_precision_integers)
> are already specified in the arguments of LLVM-to-SPIRV conversion:
> but SPIRV-gen phase does not replace them with some emulation sequences and keeps operations like this (see SPIRV converted back to LLVM-IR):

Enabling SPV_KHR_bit_instructions means that the target supports bit instructions like bitreverse, so llvm-spirv will not lower llvm_bitreverse* intrinsic calls.  If SPV_KHR_bit_instructions is disabled then llvm-spirv will transform the llvm_bitreverse* intrinsic calls.

**Comment by sarnex** [comment]
I'm taking over this work so I'll close the PR and resubmit a new one with a few changes.
